src/util/bip152.rs

```diff
@@ -0,0 +1,438 @@
+
+        // 3. Dropping the 2 most significant bytes from the SipHash output to make it 6 bytes.
+        let mut id = ShortId([0; 6]);
+        id.0.copy_from_slice(&hash[0..6]);
+        id
+    }
+}
+
+impl hex::FromHex for ShortId {
+    fn from_byte_iter<I>(iter: I) -> Result<Self, hex::Error>
+    where
+        I: Iterator<Item = Result<u8, hex::Error>> + ExactSizeIterator + DoubleEndedIterator,
+    {
+        Ok(ShortId(hex::FromHex::from_byte_iter(iter)?))
+    }
+}
+
+impl fmt::LowerHex for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        hex::format_hex(&self.0[..], f)
+    }
+}
+
+impl fmt::Display for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl fmt::Debug for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl Encodable for ShortId {
+    #[inline]
+    fn consensus_encode<S: io::Write>(&self, s: S) -> Result<usize, encode::Error> {
+        self.0.consensus_encode(s)
+    }
+}
+
+impl Decodable for ShortId {
+    #[inline]
+    fn consensus_decode<D: io::Read>(d: D) -> Result<ShortId, encode::Error> {
+        Ok(ShortId(Decodable::consensus_decode(d)?))
+    }
+}
+
+/// A [HeaderAndShortIDs] structure is used to relay a block header, the short
+/// transactions IDs used for matching already-available transactions, and a
+/// select few transactions which we expect a peer may be missing.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct HeaderAndShortIds {
+    /// The header of the block being provided.
+    pub header: BlockHeader,
+    ///  A nonce for use in short transaction ID calculations.
+    pub nonce: u64,
+    ///  The short transaction IDs calculated from the transactions
+    ///  which were not provided explicitly in prefilled_txs.
+    pub short_ids: Vec<ShortId>,
+    ///  Used to provide the coinbase transaction and a select few
+    ///  which we expect a peer may be missing.
+    pub prefilled_txs: Vec<PrefilledTransaction>,
+}
+impl_consensus_encoding!(HeaderAndShortIds, header, nonce, short_ids, prefilled_txs);
+
+impl HeaderAndShortIds {
+    /// Create a new [HeaderAndShortIds] from a full block.
+    ///
+    /// The version number must be either 1 or 2.
+    ///
+    /// The [prefill] slice indicates which transactions should be prefilled in
+    /// the block. It should contain the indexes in the block of the txs to
+    /// prefill. It must be ordered. 0 should not be included as the
+    /// coinbase tx is always prefilled.
+    ///
+    /// > Nodes SHOULD NOT use the same nonce across multiple different blocks.
+    ///
+    /// Possible [Error] variants:
+    /// - [UnknownVersion]: If the version is not 1 or 2.
+    /// - [InvalidPrefill]: If the [prefill] slice was invalid.
+    pub fn from_block(
+        block: &Block,
+        nonce: u64,
+        version: u32,
```

Comment:
Just wondering if it would be a good idea to introduce a "BlockStandardVersion" enum, in this function would remove the version check and the `unreachable()!` macro calls

Reply:
Hmm, like this, the error only happens when creating one. At least we can decode newer versions from a byte stream and then notice that it has a version we don't understand or so..